### PR TITLE
Github Actions

### PR DESCRIPTION
**Pull request by NicholasHallman**
While this PR only covers testing I'm taking this time to highlight some of the specifics of this approach. 

## Functional Testing

d2l-router
 - [x] Has a list of routes to send to page.js 

d2l-route
 - [x] Fail if there is no router parent
 - [x] Activate if it's path is selected
 - [x] Deactivate if it's not our path
 - [x] Observes the context
 - [x] Updates when the context changes

## Performance

The route runs it's connected callback first, it makes sure there is a d2l-router component somewhere as a parent.  

Then the router checks it's children for d2l-routes, When it finds one it adds it to a list of routes, takes it's path and registers it with page  js. This order is necessary as the middleware that the router initializes needs to be registerd BEFORE we add the paths.

While the router is adding paths it will also update the observers to make sure the routes have their needed contexts before first render.

Once this is finished and the router updates once it will start PageJS. 

Dev-Ops  N/A  

Cost N/A  

Documentation  

UX N/A

FYI: @anhill-D2L @Vitalii-Misechko

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      - name: Linting
+        run: npm run lint
+  sauce:
+    timeout-minutes: 15
+    runs-on: ubuntu-latest
+    steps:
+      - uses: Brightspace/third-party-actions@actions/checkout
+      - uses: Brightspace/third-party-actions@actions/setup-node
+        with:
+          node-version: 12.x
```
Comment (johngwilkinson) [inline .github/workflows/ci.yml:22]: maybe node 14, like the new AQ dashboard?

File: sc-karma-sauce-1620919062.log [hunk 2]
```diff
@@ -0,0 +1,15 @@
+2021-05-13 11:17:50.622 [32312] Sauce Connect 4.6.4, build 5309 21ce484
```
Comment (johngwilkinson) [inline sc-karma-sauce-1620919062.log:1]: Looks like this file should be .gitignore-ed

**Review by johngwilkinson — COMMENTED** [review]
Looks pretty cool. I'd suggest having some Daves review this. I think @dlockhart was looking into page.js at some point.

**Comment by NicholasHallman** [comment]
> Looks pretty cool. I'd suggest having some Daves review this. I think @dlockhart was looking into page.js at some point.

Thanks! I was planning on posting it in #web-platform for other's who might be interested to weigh in. I know that @wicken expressed some interest in a convo there about 5 months ago.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+name: Draco Insights Dashboard
```
Comment (bemailloux) [inline .github/workflows/ci.yml:1]: Oops

File: README.md [hunk 4]
```diff
@@ -11,6 +13,20 @@ npm run start
 
 This will run the demo and the components in examples/components.js
 
+### Testing
+
+```bash
+npm run test
+```
+
+The repo is based on the @open-wc template and comes with husky. Testing will run each time you commit. Failed tests may keep you from pushing. Prettier is also used to format the code. You may need to run
```
Comment (bemailloux) [inline README.md:22]: \*insert "feels good" meme here\*

File: README.md [hunk 5]
```diff
@@ -11,6 +13,20 @@ npm run start
 
 This will run the demo and the components in examples/components.js
 
+### Testing
+
+```bash
+npm run test
```
Comment (bemailloux) [inline README.md:19]: `npm test` is probably sufficient

File: src/d2l-router.js [hunk 6]
```diff
@@ -10,120 +10,120 @@ import { html, nothing } from 'lit-html';
 import page from 'page';
 
 page.observeContext = function (callback) {
-  if (!this._contextObservers) this._contextObservers = [];
-  this._contextObservers.push(callback);
+    if (!this._contextObservers) this._contextObservers = [];
+    this._contextObservers.push(callback);
```
Comment (bemailloux) [inline src/d2l-router.js:14]: Ironic that all these whitespace changes allow us to review the actual code lol

File: demo/index.html [hunk 7]
```diff
@@ -3,6 +3,7 @@
 <html>
     <head>
         <script type="module"> 
+            /* eslint-disable import/no-unresolved */
             import 'http://localhost:8000/src/d2l-router.js';
             import 'http://localhost:8000/examples/components.js';
```
Comment (bemailloux) [inline demo/index.html:8]: Wouldn't a relative path work here? In a few cases someone might want to concurrently run a few dev servers, in which case using the absolute path won't work if the port isn't 8000.

**Review by bemailloux — APPROVED** [review]

File: demo/index.html (hunk 7 above)
Comment (NicholasHallman) [reply demo/index.html:8]: Good point I'll fiddle and see if I can get that working

**Review by johngwilkinson — APPROVED** [review]

**Pull request merged**